x-pack/plugins/security_solution/public/resolver/store/middleware/resolver_tree_fetcher.ts

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+/* eslint-disable no-duplicate-imports */
+
+import { Dispatch, MiddlewareAPI } from 'redux';
+import { ResolverTree } from '../../../../common/endpoint/types';
+
+import { KibanaReactContextValue } from '../../../../../../../src/plugins/kibana_react/public';
+import { ResolverState, ResolverAction } from '../../types';
+import * as selectors from '../selectors';
+import { StartServices } from '../../../types';
+/**
+ * A function that handles syncing ResolverTree data w/ the current entity ID.
+ * This will make a request anytime the entityID changes (to something other than undefined.)
+ * If the entity ID changes while a request is in progress, the in-progress request will be cancelled.
+ * Call the returned function after each state transition.
+ * This is a factory because it is stateful and keeps that state in closure.
+ */
+export function ResolverTreeFetcher(
+  context: KibanaReactContextValue<StartServices>,
+  api: MiddlewareAPI<Dispatch<ResolverAction>, ResolverState>
+): () => void {
+  let lastRequest:
+    | { readonly abortController: AbortController; readonly entityID: string }
+    | undefined;
+
+  // Call this after each state change.
+  // This fetches the ResolverTree for the current entityID
+  // if the entityID changes while
+  return async () => {
+    const state = api.getState();
+    const entityIDToFetch = selectors.entityIDToFetch(state);
+
+    if (selectors.shouldAbortRequest(state) && lastRequest) {
+      lastRequest.abortController.abort();
+      // calling abort will cause an action to be fired
+    } else if (entityIDToFetch !== null) {
+      lastRequest = {
+        abortController: new AbortController(),
+        entityID: entityIDToFetch,
+      };
+      let result: ResolverTree | undefined;
+      // Inform the state that we've made the request. Without this, the middleware will try to make the request again
+      // immediately.
+      api.dispatch({
+        type: 'appRequestedResolverData',
+        payload: entityIDToFetch,
+      });
+      try {
+        result = await context.services.http.get(`/api/endpoint/resolver/${entityIDToFetch}`, {
+          signal: lastRequest.abortController.signal,
+          query: {
+            children: 5,
+            ancestors: 5,
+          },
+        });
+      } catch {
+        api.dispatch({
+          type: 'serverFailedToReturnResolverData',
```

Comment:
❔ This runs when a signal aborts right? Seems like we should have a distinct `catch` for `AbortError` and an action like `appReceivedAbortSignalDuringResolverDataFetch`

Reply:
seems like a good idea. i'm going to try that.

Reply:
Something like:
```
      } catch (error) {
        // https://developer.mozilla.org/en-US/docs/Web/API/DOMException#exception-AbortError
        if (error instanceof DOMException && error.name === 'AbortError') {
          api.dispatch({
            type: 'appAbortedResolverDataRequest',
            payload: databaseDocumentIDToFetch,
          });
        } else {
          api.dispatch({
            type: 'serverFailedToReturnResolverData',
            payload: databaseDocumentIDToFetch,
          });
        }
      }
```
?